File: 1114-print-in-order/1114-print-in-order.cpp
class Foo {
private:
    int counter;
    mutable std::mutex mut;
    mutable std::condition_variable cond;

public:
    Foo()
    : counter{1}
    , mut{}
    , cond{}
    {
    }

    void first(function<void()> printFirst) {
        std::unique_lock lock{mut};
        cond.wait(lock, [this](){
            return counter == 1;
        });
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        ++counter;
        lock.unlock();
        cond.notify_all();
    }

    void second(function<void()> printSecond) {
        std::unique_lock lock{mut};
        cond.wait(lock, [this](){
            return counter == 2;
        });
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        ++counter;
        lock.unlock();
        cond.notify_all();
    }

    void third(function<void()> printThird) {
        std::unique_lock lock{mut};
        cond.wait(lock, [this](){
            return counter == 3;
        });
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
        counter = 1;
        lock.unlock();
        cond.notify_all();
    }
};

File: 1114-print-in-order/1114-print-in-order.java
class Foo {

    private final ReentrantLock lock;
    private final Condition cond;
    private int counter;

    public Foo() {
        lock = new ReentrantLock();
        cond = lock.newCondition();
        counter = 1;
    }

    public void first(Runnable printFirst) throws InterruptedException {
        lock.lock();
        try {
            // printFirst.run() outputs "first". Do not change or remove this line.
            printFirst.run();
            ++counter;
            cond.signalAll();
        } finally {
            lock.unlock();
        }
    }

    public void second(Runnable printSecond) throws InterruptedException {
        lock.lock();
        try {
            while (counter < 2) {
                cond.await();
            }
            // printSecond.run() outputs "second". Do not change or remove this line.
            printSecond.run();
            ++counter;
            cond.signalAll();
        } finally {
            lock.unlock();
        }
    }

    public void third(Runnable printThird) throws InterruptedException {
        lock.lock();
        try {
            while (counter < 3) {
                cond.await();
            }
            // printThird.run() outputs "third". Do not change or remove this line.
            printThird.run();
        } finally {
            lock.unlock();
        }
    }
}